### DISCUSSION TITLE
Ikev2 uniqueness issue with unique = keep

### DISCUSSION THREAD

**Discussion by cyanideangel2000** [strongswan/strongswan#1792] [Q&A] [answered]
Hello All,

I am seeing a behavior with strongswan 5.9.7 that seems like a bug to me.
We are Ikkev2 responder and have set unique = keep.
We were expecting this will ensure no new Ike SA gets established between peers with identical peer ID.
However, what we observe is this is allows us to accept ike negotiation and we end up with multiple duplicate tunnels between the same peers.
When we switch to unique = replace, we do notice strongswan to detect the  duplication and trigger the removal of older ike established.
Therefore it appears strongswan is able to detect duplicacy however with keep set it isn't rejecting the attempt.
Is this a bug that somehow skipped our radar?

Thank you in advance.
Varun

**Comment by tobiasbrunner** [comment]
Please post the logs that show this.

**Reply by tewarivarun** [reply]
Hello @tobiasbrunner Anything specific you looking for.
I can enable higher logging level and collect them.

With unique = replace, we have the following
`schedule delete of duplicate IKE_SA for peer 'A.B.C.D' due to uniqueness policy and suspected reauthentication`

With unique = keep
swanctl --list-sas --ike ConnName
shows multiple Ike for the ConnName and all in ESTAB state
This ends up invoking our ike_updown/child_updown callback with up = true for each of these IKEs and fill up all our limits that we have set, when in my design I was expecting our callbacks to be hit just once and rest all attempts strong swan will reject and send peer IKE notify of some sort ?

**Reply by tobiasbrunner** [reply, marked as answer]
If you connect from the same IP and port (which that log excerpt suggests) then this is expected as reauthentication is explicitly allowed (and `keep` make obviously no sense in that case so the new SA is not prevented).

**Reply by tewarivarun** [reply]
Ah okay, yes, the network tuple is identical in all cases as everything between same peers on 4500 ports.
So this means, there isn't any legitimate way to differentiate between spurious connects from such weird peers vs authentic re-auth scenario.

**Reply by tobiasbrunner** [reply]
Nope, unfortunately, a reauthentication is just like a completely fresh IKE_SA.

**Reply by tewarivarun** [reply]
I see, is there any rate limiting config that strong swan has that I might have overlooked that can be set per tunnel to mitigate this from such rogue peers ?

**Reply by tobiasbrunner** [reply]
There are some thresholds for half-open SAs (triggering cookies or blocking them completely), but not for established SAs. You'd have to implement something like that yourself (e.g. via a plugin or vici client).

Also, why use `keep` and not `replace` if the latter does basically what you want?

**Reply by cyanideangel2000** [reply]
replace is our choice as well, but since replace leads to both up and down event, and our callbacks involve setting up few interfaces and other fds, too many flaps first spike udevd on ubuntu and then leads to lots of epoll ctl fd activiity and related things. It gets messy, therefore was experimenting with keep and didn't see the behavior that i was expecting.
Yep, I would need to have some ratelimiting in place in our callbacks to protect against such rogue peers.

**Reply by tobiasbrunner** [reply]
> but since replace leads to both up and down event,

Which will also be the case with regular make-before-break reauthentication. With `keep` you'll also prevent clients from reconnecting when they e.g. lost connectivity and restarted the connection without sending a delete to the server.

**Reply by cyanideangel2000** [reply]
Ack'ed. Thank you @tobiasbrunner for having this discussion here.
